javalib/src/main/scala/java/net/URI.scala

```diff
@@ -107,16 +107,22 @@ final class URI(origStr: String) extends Serializable with Comparable[URI] {
       else x.fold(-1)(s1 => y.fold(1)(s2 => cmp(s1, s2)))
     }
 
-    if (this._scheme != that._scheme)
-      this._scheme.fold(-1)(s1 => that._scheme.fold(1)(s1.compareToIgnoreCase))
-    else if (this._isOpaque != that._isOpaque)
-      // A hierarchical URI is less than an opaque URI
-      if (this._isOpaque) 1 else -1
-    else if (_isOpaque) {
+    @inline def cmpSsp(): Int =  {
       val ssp = cmp(this._schemeSpecificPart, that._schemeSpecificPart)
       if (ssp != 0) ssp
       else cmpOpt(this._fragment, that._fragment)
-    } else if (this._authority != that._authority) {
+    }
+
+    if (this._scheme != that._scheme) {
+      val s = this._scheme.fold(-1)(s1 => that._scheme.fold(1)(s1.compareToIgnoreCase))
+      if (s == 0 && isOpaque()) cmpSsp()
```

Comment:
I'm not sure this now works properly if two URIs have the same scheme but one is opaque an the other is not.

Consider replacing all of this with something like this:

```scala
List(
  this._scheme.fold(-1)(s1 => that._scheme.fold(1)(s1.compareToIgnoreCase)),
  this._isOpaque.toInt - that._isOpaque.toInt, // A hierarchical URI is less than an opaque URI
  ...
).find(_ != 0).getOrElse(0)
```

This makes it unmistakably clear that these are merely priorities for comparison. Also, it will fix similar bugs in other parts when the supplied `cmp` is used.

Reply:
Refactored in https://github.com/scala-js/scala-js/pull/3690/commits/b2caa31c242ad85efee3fe748e9a7b925f220a0c and fixed an existing issue in subequent https://github.com/scala-js/scala-js/pull/3690/commits/52d33ceece6e3fb4b655a1ccc0dbf7c4df2fe7d7